packages/contract/src/EphemeralNotes.sol

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.12;
+
+import "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import "openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol";
+
+struct Note {
+    address ephemeralOwner;
+    address from;
+    uint256 amount;
+}
+
+contract EphemeralNotes {
+  mapping(address => Note) public notes;
+  IERC20 public immutable token;
+  
+  event NoteCreated(Note note);
+  event NoteRedeemed(Note note);
+
+  constructor(IERC20 _token) {
+    token = _token;
+  }
+
+
+  // Call token.approve(<address of this contract>, amount) on the token contract before this
```

Comment:
we can unlimited approve a few trusted contracts on account creation
- this one, for notes
- the paymaster, to pay fees in USDC